Fix & improve Lucas-Kanade optical flow

pxl-th opened the pull request:
Hi! I was trying to use LK optical flow and noticed that it didn't give the correct results.
Either the flow was too big or its direction was incorrect. Additionally it was dropping a lot of points, marking them as bad.

#### Bugs:

Upon investigation I've noticed two major bugs:

1. Incorrect pyramid level in the `determine_pyramid_coordinates` function.
    The indexing starts at `1`, but when computing pyramid point, it should start at `0`, so the first level does not change the coordinates, since it is the original level.
2. In the beginning of the optical flow there is the conversion of points and initial displacements from `(row, col)` to `(x, y)` format.
    However, all subsequent functions operate in the `(row, col)` format.
    Simple way to check this is to visualize the gradient windows for the points.

#### Changes:

Do tell me if you don't agree with some of the changes.

- Use the correct index for pyramid level.
- Drop the conversion from `(row, col)` to `(x, y)` and operate only in `(row, col)` format.
- Set default value for the `eigenvalue_threshold` parameter to `1e-4` as in [OpenCV](https://docs.opencv.org/4.5.2/dc/d6b/group__video__track.html#ga473e4b886d0bcc6b65831eb88ed93323).
- Remove `in_image` function, since `lies_in` is enough.
- Convert `while` loops to explicit expressions in `get_grid` function.
- Simplify displacement creation in tests.
- Use default parameters in tests for Lucas-Kanade algorithm.
- Refactoring.
  Replace tabs in certain places with spaces for consistency purposes.
  Make rows no more than 92 characters long & minor refactoring (this allows for split-screen editing 🙃).
- Add epsilon termination criteria as in [OpenCV LK](https://docs.opencv.org/4.5.2/d9/d5d/classcv_1_1TermCriteria.html#a56fecdc291ccaba8aad27d67ccf72c57a857609e73e7028e638d2ea649f3b45d5).
- Update LK tests to require bigger accuracy.
- Remove `is_truncated_window`. Use `get_grid` to determine whether we need to recompute gradient or can use the previous one.
- Remove `flow` vector, since it was used only to store pyramid layer contributions and was redundant.
- Remove extrapolation and use only interpolation for the second pyramid layers.
  Since we never operate on the out-of-bounds regions.
- Use svd2x2 to compute minimum eigenvalues instead of calling `eigen`.
- Make `optflow!` function operate on the `LKPyramid` structure that holds image pyramid and gradients.

#### TODOs:
- [ ] Add ability to pass precomputed pyramids and image gradients.
      This can be useful, when computing flow for the image with more than one other image.
      Since this way we would need to compute them only once.
      I think about either adding them to keyword parameters, or creating `LKPyramid` struct that will contain both pyramids and gradients. Creating struct will decrease number of parameters in the function, which would make things easier.

#### Visual comparison:

I've used the same code for visualization as in the example docs.

Starting with the simple case, where the figure moves only horizontally.

|Current|This PR|
|:-:|:-:|
|![lk-old-1](https://user-images.githubusercontent.com/17990405/124880432-f4eb7880-dfd6-11eb-8a7f-5fe88122371a.png)|![lk-new-1](https://user-images.githubusercontent.com/17990405/124880455-f9b02c80-dfd6-11eb-8d02-8015fb5e6afb.png)|

Car moves forward. Two adjacent frames in a video.

|Current|This PR|
|:-:|:-:|
|![lk-old-3](https://user-images.githubusercontent.com/17990405/124880624-2b28f800-dfd7-11eb-9c4b-8dafa97bed83.png)|![lk-new-3](https://user-images.githubusercontent.com/17990405/125197376-0ac29d00-e266-11eb-9153-911ec6757174.png)|

Basketball example. Motion is mainly in the hands and the body. Static background.

|Current|This PR|
|:-:|:-:|
|![lk-old-4](https://user-images.githubusercontent.com/17990405/124880847-61667780-dfd7-11eb-9e55-37676d1c9788.png)|![lk-new-4](https://user-images.githubusercontent.com/17990405/125197394-17df8c00-e266-11eb-8d77-b64f6fcb07b3.png)|

Cup. Horizontal movement.

|Current|This PR|
|:-:|:-:|
|![lk-old-2](https://user-images.githubusercontent.com/17990405/124881324-dfc31980-dfd7-11eb-83bb-2f4bc863afea.png)|![lk-new-2](https://user-images.githubusercontent.com/17990405/125197415-27f76b80-e266-11eb-9c4e-1cf7a8212e98.png)|

Commit:
Not complete fix

Commit:
Finish fixing LK optical flow

Commit:
Refactor code

Simplify displacement creation in tests.

zygmuntszpak commented:
Thanks for working on this. Looks very promising! I'll need a couple of days to reacquaint myself with the code and algorithm. There is definitely a problem with the current implementation because I can see in the old code that we have ``Ix, Iy = imgradients``  instead of ``Iy, Ix = imgradients`` which is one of the several things you have fixed!

zygmuntszpak commented:
> Hi! I was trying to use LK optical flow and noticed that it didn't give the correct results.
> Either the flow was too big or its direction was incorrect. Additionally it was dropping a lot of points, marking them as bad.
> #### Bugs:
> 
> Upon investigation I've noticed two major bugs:
> 
>     1. Incorrect pyramid level in the `determine_pyramid_coordinates` function.
>        The indexing starts at `1`, but when computing pyramid point, it should start at `0`, so the first level does not change the coordinates, since it is the original level.
> 
>     2. In the beginning of the optical flow there is the conversion of points and initial displacements from `(row, col)` to `(x, y)` format.
>        However, all subsequent functions operate in the `(row, col)` format.
>        Simple way to check this is to visualize the gradient windows for the points.

Thanks for addressing these colossal bugs. 

> 
> #### Fixes:
> 
> Do tell me if you don't agree with some of the changes.
> 
>     * Use the correct index for pyramid level.
> 
>     * Drop the conversion from `(row, col)` to `(x, y)` and operate only in `(row, col)` format.
> 
>     * Set default value for the `eigenvalue_threshold` parameter to `1e-4` as in [OpenCV](https://docs.opencv.org/4.5.2/dc/d6b/group__video__track.html#ga473e4b886d0bcc6b65831eb88ed93323).
> 
>     * Remove `in_image` function, since `lies_in` is enough.
> 
>     * Convert `while` loops to explicit expressions in `get_grid` function.
> 
>     * Simplify displacement creation in tests.
> 
>     * Use default parameters in tests for Lucas-Kanade algorithm.
> 
>     * Replace tabs in certain places with spaces for consistency purposes.
> 
>     * Make rows no more than 80 characters long & minor refactoring (this allows for split-screen editing upside_down_face).

I think overall these are some excellent changes and I am grateful for the time you put into making them. 

If we are going to be changing the code style then I suggest we follow the [BlueStyle](https://github.com/invenia/BlueStyle) which I believe already has a lot of traction in the Julia community. One of the key implications is that the 80 character limit would become a 92 character limit. 



> 
> #### TODOs:
> 
>     * [ ]  Add epsilon termination criteria as in [OpenCV LK](https://docs.opencv.org/4.5.2/d9/d5d/classcv_1_1TermCriteria.html#a56fecdc291ccaba8aad27d67ccf72c57a857609e73e7028e638d2ea649f3b45d5).
> 
>     * [ ]  Add ability to pass precomputed pyramids and image gradients.
>       This can be useful, when computing flow for the image with more than one other image.
>       Since this way we would need to compute them only once.
>       I think about either adding them to keyword parameters, or creating `LKPyramid` struct that will contain both pyramids and gradients. Creating struct will decrease number of parameters in the function, which would make things easier.
> 
>     * [ ]  Add tests that fail the current LK version, but succeed on this PR.
> 
>     * [ ]  Bring back deleted comments (this was done for simplification).
> 
> 

Those are great suggestions. However, I propose you keep the "ability to pass precomputed pyramids and image gradients." for a separate pull request. 

One important issue that needs to be addressed in this PR is an apparent substantial increase in runtime. Using [BenchmarkTools](https://github.com/JuliaCI/BenchmarkTools.jl) I found that the existing code takes approximately 1 second to complete on a pair of test images on my machine, whereas this pull-request takes 5 seconds for the same pair. I believe the issue is that your proposed version allocates memory within performance-critical loops. I think this can be traced down to you removing the use of the custom pinv2x2 function in favour of the built-in pinv function. Revisting this code, I think that we may be better-off using the svd2x2 function directly. This will allow us to construct the pseudo-inverse, but also give us the singular values which we can use to define min_eigenvalue (if I recall correctly, for real symmetric matrices with non-negative eigenvalues, the eigenvalues and singular values are the same). 

Thanks again for reviving this implementation.

zygmuntszpak commented on src/lucas_kanade.jl:
```diff
@@ -1,259 +1,239 @@
+function get_pyramid_coordinates(points, i, j)
+    px = floor(Int, points[j][1] / 2 ^ (i - 1))
+    py = floor(Int, points[j][2] / 2 ^ (i - 1))
+    SVector{2}(px ,py)
```
In light of using the row/column convention, shouldn't this be
```julia
   py = floor(Int, points[j][1] / 2 ^ (i - 1))
   px = floor(Int, points[j][2] / 2 ^ (i - 1))
   SVector{2}(py ,px)
```

Commit:
Add LKPyramid support

- Remove `is_truncated_window`. Use `get_grid` to determine whether we
  need to recompute gradient or can use the previous one.
- Remove `flow` vector, since it was used only to store pyramid layer
  contributions and was redundant.
- Remove extrapolation and use only interpolation for the second pyramid
  layers. Since we never operate on the out-of-bounds regions.
- Use svd2x2 to compute minimum eigenvalues instead of calling `eigen`.
- Update LK tests to require bigger accuracy.

Commit:
Change epsilon termination criteria

pxl-th commented:
Hi! Thanks for the review. I've made several improvements and changes based on your comments.

One thing to note is that at the moment of your review I've already implemented version of the algorithm that operates on `LKPyramid` struct that holds pyramid for the image and its gradients.
However, I've not exposed the functionality to the user, so from API standpoint it is still the same.
Only underlying functions are utilizing `LKPyramid` struct.
I'd appreciate if we can add these changes in this pull request. But we can expose the necessary API in a separate PR.

I've also used `svd2x2` to compute minimum singular values, since they indeed coincide with the eigenvalues in this case.

#### Quick summary of changes from previous review
- Changed max line width to 92 symbols.
- Remove `is_truncated_window`.
  Use `get_grid` to determine whether we need to recompute gradient or can use the previous one.
  Since `is_truncated_window` always required to recompute the gradients, but between iterations the grid can stay the same (even with the truncated window), leading to redundant recomputations.
- Remove `flow` vector, since it was used only to store pyramid layer contributions and was redundant.
- Remove extrapolation and use only interpolation for the second pyramid's layers.
  Since we never operate on the out-of-bounds regions.
- Use `svd2x2` to compute minimum eigenvalues instead of calling `eigen` (singular values conicide with eigenvalues in this case).
- Add epsilon termination criteria. Add the necessary parameter to `LucasKanade` struct as well.
- Update LK tests to require higher accuracy.

#### Benchmarking results (on 1000 initial keypoints)

Current:
`1.555 s (866161 allocations: 260.38 MiB)`

This PR:
- With epsilon termination criteria (used by default): `762.053 ms (7079 allocations: 240.53 MiB)`
- Without epsilon termination criteria: `1.827 s (7091 allocations: 240.53 MiB)`
- With precomputed pyramids: `301.647 ms (10 allocations: 10.31 MiB)`

Note that without epsilon termination criteria the time is slightly bigger.
This can be because current version and this PR are tracking and invalidating different amount of points.

Commit:
Move epsilon checking to after out-of-bounds checking

johnnychen94 commented on src/core.jl:
```diff
@@ -2,7 +2,7 @@
 # Blinn, J. (1996). Consider the lowly 2 x 2 matrix. IEEE Computer Graphics and Applications, 16(2), 82-88.
 # https://scicomp.stackexchange.com/questions/8899/robust-algorithm-for-2-times-2-svd
-function svd2x2(M::AbstractArray)
+function svd2x2(M::AbstractMatrix{T}) where T <: Number
```
For performance consideration, can we make the input `M` a SMatrix?

pxl-th replied:
Farneback algorithm uses matrix views with this function.
I guess it can be changed in a separate PR if needed, to keep this PR from growing in size even more.

johnnychen94 replied:
Never mind, I just realize that it actually makes no differences at all:

```julia
julia> @btime svd2x2($(SMatrix{2,2}(rand(2, 2))));
  43.171 ns (0 allocations: 0 bytes)

julia> @btime svd2x2($(rand(2, 2)));
  43.698 ns (0 allocations: 0 bytes)
```

johnnychen94 commented on src/optical_flow.jl:
```diff
@@ -102,25 +102,36 @@ a default value of four is assumed.
 
-LucasKanade(iterations::Int = 20; window_size::Int = 11, pyramid_levels::Int = 4, eigenvalue_threshold::Real = 0.000001) = LucasKanade(iterations, window_size,  pyramid_levels, eigenvalue_threshold)
+LucasKanade(
+    iterations::Int = 20; window_size::Int = 11, pyramid_levels::Int = 4,
```
This type restriction is usually unnecessary. The constructor would implicitly call `convert(Int, iterations)`.

```suggestion
    iterations::Integer = 20; window_size::Integer = 11, pyramid_levels::Integer = 4,
```

johnnychen94 commented on src/optical_flow.jl:
```diff
@@ -102,25 +102,36 @@ a default value of four is assumed.
     pyramid_levels::I
     eigenvalue_threshold::F
+    ϵ::Float64
 end
```
No parameters actually.
```julia
struct LucasKanade <: OpticalFlowAlgorithm
    iterations::Int
    window_size::Int
    pyramid_levels::Int
    eigenvalue_threshold::Float64
    ϵ::Float64
end
```

pxl-th replied:
Sorry, didn't get what you mean here. Do you suggest removing this parameter?

pxl-th replied:
Nevermind, got it :)

johnnychen94 commented on test/lucas_kanade.jl:
```diff
@@ -1,151 +1,125 @@
-# ten-minute timeout limit on Travis. Travis assumes the tests have hung
-# if the interval between printing something to stdio exceeds ten minutes.
 using Images
-@info "Finished loading Images package."
```
A good change 👍 . This kind of information is usually over verbose.

johnnychen94 commented on test/lucas_kanade.jl:
```diff
@@ -1,151 +1,125 @@
 
+function get_keypoints(image, n_points)
+    corners = imcorner(image, method=shi_tomasi)
+    I = findall(!iszero, corners)
```
`findall` uses a for loop version, we can instead use broadcasting here, which is usually faster:

```suggestion
    I = LinearIndices(image)[corners .!≈ 0]
```

```julia
X = rand(64, 64);
J = randperm(prod(size(X)))[1:64];
X[J] .= 0;

I1 = findall(!iszero, X);
I2 = CartesianIndices(X)[X .!= 0];
I1 == I2 # true

@btime findall(!iszero, $X); # 34.101 μs (13 allocations: 128.61 KiB)
@btime CartesianIndices($X)[$X .!= 0] # 17.237 μs (5 allocations: 67.92 KiB)
```

---

Edit: oops this is test codes. Well, feel free to ignore this suggestion. But do keep in mind the differences.

johnnychen94 commented on src/lucas_kanade.jl:
```diff
@@ -1,259 +1,240 @@
+    points::Vector{SVector{2, Float64}}, displacement::Vector{SVector{2, Float64}},
+    algorithm::LucasKanade,
+)
+    enough_layers = (
+        length(first_pyramid.layers) > algorithm.pyramid_levels &&
+        length(second_pyramid.layers) > algorithm.pyramid_levels
+    )
+    !enough_layers && throw("Not enough layers in pyramids.")
```
A style suggestion and feel free to ignore.

```suggestion
    has_enough_layers = length(first_pyramid.layers) > algorithm.pyramid_levels &&
        length(second_pyramid.layers) > algorithm.pyramid_levels
    has_enough_layers || throw("Not enough layers in pyramids.")
```

johnnychen94 commented on src/lucas_kanade.jl:
```diff
@@ -1,259 +1,240 @@
+    )
+    !enough_layers && throw("Not enough layers in pyramids.")
+
+    n_points = points |> length
```
It is not recommended to have `|>` in library codes.

```suggestion
    n_points = length(points)
```

johnnychen94 commented on src/lucas_kanade.jl:
```diff
@@ -1,259 +1,240 @@
-    else
-        val = min_eigenvalue/window
+function compute_partial_derivatives(Iy, Ix)
+    Iyy = imfilter(Iy .* Iy, Kernel.gaussian(4))
+    Ixx = imfilter(Ix .* Ix, Kernel.gaussian(4))
+    Iyx = imfilter(Iy .* Ix, Kernel.gaussian(4))
```
Here we can use `imfilter!` to reuse the allocated memory, and use [factored kernels](https://juliaimages.org/ImageFiltering.jl/dev/#Factored-kernels) for more efficient filtering.

pxl-th replied:
Thanks for the suggestion! I'm now reusing arrays and using explicitly factored kernels.
However, turns out, from the documentation `If the kernel is a two dimensional array, imfilter will attempt to factor it; if successful, it will use the separable algorithm.`. Thus I didn't notice much of the performance improvements, but there was a slight decrease in the number of allocations. Reusing array improved performance quite a bit though.

johnnychen94 commented on src/core.jl:
```diff
@@ -16,17 +16,30 @@ function svd2x2(M::AbstractArray)
+# Computes the Moore-Penrose pseudoinverse for a 2-by-2 matrix M by only inverting
+# singular values above the threshold `tol = sqrt(eps(real(float(one(eltype(M))))))`
+function pinv2x2(M)
+    U, S, V = M |> svd2x2
```
```suggestion
    U, S, V = svd2x2(M)
```

johnnychen94 commented on src/lucas_kanade.jl:
```diff
@@ -1,259 +1,240 @@
+    G = _compute_spatial_gradient(
+        grid, pyramid.Iyy[level], pyramid.Iyx[level], pyramid.Ixx[level],
+    )
+    U, S, V = G |> svd2x2
```
```suggestion
    U, S, V = svd2x2(G)
```

johnnychen94 commented on src/lucas_kanade.jl:
```diff
@@ -1,259 +1,240 @@
+# Arguments
+- `level`: Level of the pyramid in `[1, levels]` range.
+"""
+@inline get_pyramid_coordinate(point, level) = floor.(Int64, point ./ 2 ^ (level - 1))
```
Unless there's a real reason, let's use `Int` here:

```suggestion
@inline get_pyramid_coordinate(point, level) = floor.(Int, point ./ 2 ^ (level - 1))
```

johnnychen94 commented on src/lucas_kanade.jl:
```diff
@@ -1,259 +1,240 @@
+function get_grid(point, new_point, window, image_axes)
+    rows, cols = image_axes
+
+    up = floor(Int64, min(window, min(point[1], new_point[1]) - first(rows)))
+    down = floor(Int64, min(window, last(rows) - max(point[1], new_point[1])))
+    left = floor(Int64, min(window, min(point[2], new_point[2]) - first(cols)))
+    right = floor(Int64, min(window, last(cols) - max(point[2], new_point[2])))
```
```suggestion
    up = floor(Int, min(window, min(point[1], new_point[1]) - first(rows)))
    down = floor(Int, min(window, last(rows) - max(point[1], new_point[1])))
    left = floor(Int, min(window, min(point[2], new_point[2]) - first(cols)))
    right = floor(Int, min(window, last(cols) - max(point[2], new_point[2])))
```

johnnychen94 reviewed (commented):
I don't know much about this algorithm ,so just some easy-to-handle style and performance comments.

I do notice that there're a lot of places that memory allocation can be reused by introducing in-place functions, e.g., `get_grid`, `compute_spatial_gradient`. But since I didn't profile it, I'm not sure if this is performant hot spots.

Commit:
Compute offsets separately from grid

And compute new grid only if the offsets change.
Do epsilon termination criteria before updating pyramid contribution.

Commit:
Compute partial gradients reusing allocated array

Commit:
Use KernelFactors for image filtering

Commit:
Minor refactoring

pxl-th commented:
#### Benchmarking results with the latest changes

Parameters used:
- initial number of keypoints = 1000
- window size = 21
- levels = 3
- iterations = 30

Current:
`1.555 s (866161 allocations: 260.38 MiB)`

This PR:
- With epsilon termination criteria: `352.698 ms (6839 allocations: 199.00 MiB)`
- Using precomputed pyramids: `131.568 ms (10 allocations: 10.31 MiB)`

Commit:
Remove parametrization from LucasKanade struct

Commit:
Change findall to broadcasting

johnnychen94 commented:
5x-10x is a big improvement!

zygmuntszpak commented on test/lucas_kanade.jl:
```diff
@@ -1,151 +1,125 @@
+            error_count += 1
         end
-        error = sqrt( (δ[1] - Δ[1])^2 + (δ[2] - Δ[2])^2 )
+        error = sqrt((δ[1] - Δ[1]) ^ 2 + (δ[2] - Δ[2]) ^ 2)
```
I wouldn't add those whitespaces there. I'd prefer to leave it as it was.

pxl-th replied:
Reverted

zygmuntszpak commented on src/lucas_kanade.jl:
```diff
@@ -1,259 +1,255 @@
+        # We never go out-of-bound, so there is no need to extrapolate.
+        interploated_layer = interpolate(second_pyramid.layers[level], BSpline(Linear()))
+
+        for did in 1:n_points
```
Is there a particular reason why you called the loop variable `did`? I keep reading it as the past tense of "to do" and it feels a little odd. Perhaps a vanilla `n` would work better here?

pxl-th replied:
It originally standed for `displacement_id`, but I agree, it is confusing. Changed to `n`.

zygmuntszpak reviewed (commented):
Thank you very much for your excellent work. I've left just a small remark and question, but other than that I reckon it is good to go!

Commit:
Minor refactoring

johnnychen94 reviewed (approved):
No more comments from my side. I'll leave it to @zygmuntszpak to make the decision.

johnnychen94 commented:
@pxl-th out of curiosity, there're some legacy pending PRs in this repo, are you interested in taking the role and finishing them?
